Changed output window to be a tab based window.

ChrisMarsh82 opened the pull request:
I have added R-Script as it is needed to test these changes however that work is done under  #6953

@dannyparsons Once you have a chance to look through this it would be good to decide how the main output tab should work. It is set up to be a log at the moment i.e. can't change anything. This makes sense in the way the OutputLogger is designed. If this needs to change may need to re-think the structure. Maybe another option would be to decide which tabs are the active recorders? i.e. which tabs will receive the new outputs?

lloyddewit commented:
@ChrisMarsh82 As I mentioned yesterday in PR #6953 , I made a new RScript NuGet package that does not include the XUnit dependencies (v. 1.0.4). This should reduce the number of dependencies from 32 to 5.

I suggest that you make a new PR with RScript 1.0.4, this will reduce the number of files in the PR and make it more manageable. You can then close both this PR and PR #6953 .

lloyddewit commented:
@ChrisMarsh82 I had a quick look through the files. If you make the changes above, then there are about 13 VB files for me to review.
If you can make the new PR this week, then I can aim to finish my review by mid next week. Would that be OK?

dannyparsons commented:
Looks good. I haven't found any broken or missing functionality. I've tried different kinds of output and didn't find any issues.

@africanmathsinitiative/developers We need more reviews on this as it's a big change. Particularly to check for any breaking changes.

@rdstern What do you think the name of the main tab should be? Maybe `Main`? I don't think `Output` is right.

A few comments:

1. Does there have to be scroll bars for the blocks? It makes it a bit harder to navigate and also reduces the space on the screen. I would be happy just to have the long output without the additional scroll bars.

![image](https://user-images.githubusercontent.com/11214914/142779203-4181d5a5-9732-4332-b688-56d7c5b80786.png)

2. After running a command it should scroll to the bottom. It doesn't currently.
3. The sizes seem quite different in this new output than the previous one. The output was a lot bigger than before. Not a problem, but we may need to adjust the default sizes for comments, code and output. 
4. The output window seems to refresh when a dialog is closed without running a command. It should only refresh when a command has been run.
5. I would like to be able to delete an item from the Output tab. For example, the first output I made was way too big and didn't display nicely and make the whole window a bit messy, so I wanted to delete it and do it again.
6. There used to be a right click menu in the Output Window. Some of these options will still be needed.
![image](https://user-images.githubusercontent.com/11214914/142779998-1897648a-1451-4a0b-96a0-85efc30d11ba.png)
7. `Ctrl + C` when highlighting text doesn't seem to work anymore.
**Other comments that are non essential and could be addressed in future work**
8. When I saved the output as an RFT it was 96MB but it only had one graph in and some text output. That seems quite big. Not sure if that was happening before too. Not a serious problem but the file is bigger than I expected.
9. The multiple tabs works well. It would be good to be able to delete a secondary tab directly. I'm not sure the tab should be deleted when all the items are deleted - can be discussed.
10. It would be nice to be able to scroll up and down the window with the mouse wheel, or keyboard up/down arrows.
11. Select/Deselect all buttons would be nice.

N-thony commented:
> Looks good. I haven't found any broken or missing functionality. I've tried different kinds of output and didn't find any issues.
> 
> @africanmathsinitiative/developers We need more reviews on this as it's a big change. Particularly to check for any breaking changes.
> 
> @rdstern What do you think the name of the main tab should be? Maybe `Main`? I don't think `Output` is right.
> 
> A few comments:
> 
> 1. Does there have to be scroll bars for the blocks? It makes it a bit harder to navigate and also reduces the space on the screen. I would be happy just to have the long output without the additional scroll bars.
> 
> ![image](https://user-images.githubusercontent.com/11214914/142779203-4181d5a5-9732-4332-b688-56d7c5b80786.png)
> 
> 2. After running a command it should scroll to the bottom. It doesn't currently.
> 3. The sizes seem quite different in this new output than the previous one. The output was a lot bigger than before. Not a problem, but we may need to adjust the default sizes for comments, code and output.
> 4. The output window seems to refresh when a dialog is closed without running a command. It should only refresh when a command has been run.
> 5. I would like to be able to delete an item from the Output tab. For example, the first output I made was way too big and didn't display nicely and make the whole window a bit messy, so I wanted to delete it and do it again.
> 6. There used to be a right click menu in the Output Window. Some of these options will still be needed.
>    ![image](https://user-images.githubusercontent.com/11214914/142779998-1897648a-1451-4a0b-96a0-85efc30d11ba.png)
> 7. `Ctrl + C` when highlighting text doesn't seem to work anymore.
>    **Other comments that are non essential and could be addressed in future work**
> 8. When I saved the output as an RFT it was 96MB but it only had one graph in and some text output. That seems quite big. Not sure if that was happening before too. Not a serious problem but the file is bigger than I expected.
> 9. The multiple tabs works well. It would be good to be able to delete a secondary tab directly. I'm not sure the tab should be deleted when all the items are deleted - can be discussed.
> 10. It would be nice to be able to scroll up and down the window with the mouse wheel, or keyboard up/down arrows.
> 11. Select/Deselect all buttons would be nice.

12. I have tried to clear the output window from Tools > Clear Output Window and it seems to not work yet
     
![image](https://user-images.githubusercontent.com/68591383/142839393-0e1c324b-60c0-4576-a4e4-f1f7319ecf05.png)

dannyparsons commented:
@N-thony have you added comments here? I just see a quote of my comment.

N-thony commented:
> @N-thony have you added comments here? I just see a quote of my comment.

yes

dannyparsons commented:
I don't see your comments, can you repost them?

N-thony commented:
12. I have tried to clear the output window from Tools > Clear Output Window and it seems to not work yet
     
![image](https://user-images.githubusercontent.com/68591383/142839393-0e1c324b-60c0-4576-a4e4-f1f7319ecf05.png)

dannyparsons commented:
Thanks, sorry I didn't see that before.

lloyddewit commented:
> @ChrisMarsh82 As I mentioned yesterday in PR #6953 , I made a new RScript NuGet package that does not include the XUnit dependencies (v. 1.0.4). This should reduce the number of RScript dependencies from 32 to 5.
> 
> I suggest that you make a new PR with RScript 1.0.4, this will reduce the number of files in the PR and make it more manageable. You can then close both this PR and PR #6953 .

@ChrisMarsh82 Please could you close this PR and make a new PR without the XUnit dependencies? If you do this, then I can peer review the code in the new PR, thanks

ChrisMarsh82 commented:
@lloyddewit all dependencies are taken out now

lloyddewit commented on instat/Model/Output/clsOutputLogger.vb:
```diff
@@ -0,0 +1,179 @@
+    Public Function GetFilteredList(strListName As String, Optional bCreate As Boolean = True) As clsOutputList
+        Dim filteredList As clsOutputList
+        filteredList = _filteredOutputs.Where(Function(x) x.StrName = strListName).FirstOrDefault()
+        If filteredList Is Nothing And bCreate Then
```
Will the line above always set `filteredList` to a a found list or a default list? Will `filteredList` ever be `Nothing`?

ChrisMarsh82 replied:
It calls GetFilteredList if it exists or not. The optional bCreate decides if it should be created or return nothing if it doesn't exist

lloyddewit commented on instat/Model/Output/clsOutputLogger.vb:
```diff
@@ -0,0 +1,179 @@
+    ''' <param name="strListName"></param>
+    ''' <returns></returns>
+    Public Function IsValidFilteredListName(strListName As String) As Boolean
+        If Trim(strListName) Is "" Then
+            Return False
+        End If
+        If _filteredOutputs.Where(Function(x) x.StrName = strListName).Any() Then
```
```suggestion
        If Trim(strListName) Is "" OrElse
            _filteredOutputs.Where(Function(x) x.StrName = strListName).Any() Then
```

lloyddewit commented on instat/My Project/app.manifest:
```diff
@@ -0,0 +1,79 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
why is this file needed in this PR? why didn't R-Instat require it before?

ChrisMarsh82 replied:
removed

ChrisMarsh82 replied:
removed

lloyddewit commented on instat/UserControl/ucrOutputPage.vb:
```diff
@@ -0,0 +1,283 @@
+    ''' <returns></returns>
+    Public ReadOnly Property SelectedElements() As List(Of clsOutputElement)
+        Get
+            Dim elemets As New List(Of clsOutputElement)
```
```suggestion
            Dim elements As New List(Of clsOutputElement)
```

ChrisMarsh82 replied:
Changed

lloyddewit commented on instat/UserControl/ucrOutputPage.vb:
```diff
@@ -0,0 +1,283 @@
+    End Sub
+
+    ''' <summary>
+    ''' Copys selected elenments to clipboard
```
```suggestion
    ''' Copies selected elenments to clipboard
```

lloyddewit reviewed (changes requested):
@ChrisMarsh82 this looks good, I like the way you implemented this. Here are some initial comments.

lloyddewit commented on instat/UserControl/ucrOutputPages.vb:
```diff
@@ -0,0 +1,275 @@
+    ''' ToDo InstatOptions should be able to be accessed from anywhere
+    ''' </summary>
+    Public WriteOnly Property clsInstatOptions() As InstatOptions
+        Set(ByVal value As InstatOptions)
```
what is the purpose of 'ByVal' here?

ChrisMarsh82 replied:
This is added automatically by Visual Studio. Then visual studio tells you it's not needed. Will take it out.

lloyddewit commented on instat/UserControl/ucrOutputPages.vb:
```diff
@@ -0,0 +1,275 @@
+        Dim tabName As String = SelectedTab()
+        tabControl.TabPages.Remove(tab)
+        tab.Dispose()
+        Dim found As ucrOutputPage = Nothing
+        For Each outputPage In _allOutputPages
+            If outputPage.Tag = tabName Then
+                found = outputPage
+                Exit For
+            End If
+        Next
+        _allOutputPages.Remove(found)
```
```suggestion
        For Each outputPage In _allOutputPages
            If outputPage.Tag = tabName Then
                _allOutputPages.Remove(outputPage)
                Exit For
            End If
        Next
```

lloyddewit commented on instat/UserControl/ucrOutputPages.vb:
```diff
@@ -0,0 +1,275 @@
+    Private Sub tbAddToNew_Click(sender As Object, e As EventArgs) Handles tbAddToNew.Click
+        Dim strTabName As String = "New Tab"
+        While Not _outputLogger.IsValidFilteredListName(strTabName)
+            strTabName = strTabName + "1"
```
```suggestion
            strTabName &= "1"
```
This line will make names like "tabName1", "tabName11", "tabName111" etc.
Would "tabName1", "tabName2" etc be better?

lloyddewit commented on instat/UserControl/ucrOutputPages.vb:
```diff
@@ -0,0 +1,275 @@
+    End Sub
+
+    Private Sub tbMoveUp_Click(sender As Object, e As EventArgs) Handles tbMoveUp.Click
+        For Each element In _selectedOutputPage.SelectedElements
```
Should this explicitly specify order by ascending ID (to be consistent with line 230 above)?

ChrisMarsh82 replied:
Added in

lloyddewit commented on instat/Translations.vb:
```diff
@@ -58,7 +58,6 @@ Public Class Translations
 
         'The right mouse button menus for the 6 output windows are not accessible via 
         '    the control lists. Therefore, translate these menus explicitly
-        HandleError(TranslateWinForm.clsTranslateWinForm.TranslateMenuItems(frmMain.ucrOutput.Name, frmMain.ucrOutput.mnuContextRTB.Items, strDbPath, strLanguageCode))
```
Where are the new output windows translated?

ChrisMarsh82 replied:
Added translations back in

lloyddewit commented on instat/Translations.vb:
```diff
@@ -127,7 +126,7 @@ Public Class Translations
     '''--------------------------------------------------------------------------------------------
     Private Shared Function GetDbPath() As String
-        Dim strTranslationsPath As String = String.Concat(System.Windows.Forms.Application.StartupPath, "\translations")
+        Dim strTranslationsPath As String = String.Concat(System.Windows.Forms.Application.StartupPath, "/translations")
```
I'm assuming that `/` works with both Windows and Linux, is that correct?

ChrisMarsh82 replied:
yep

lloyddewit reviewed (changes requested):
I finished my review

ChrisMarsh82 commented on instat/Model/Output/clsOutputList.vb:
```diff
@@ -0,0 +1,81 @@
+    ''' </summary>
+    ''' <param name="outputElement"></param>
+    Public Sub MoveElementDown(outputElement As clsOutputElement)
+        If outputElement.Id < _output.Count Then
```
They do start at 1
outputElement.Id = filteredList.Output.Count + 1

ChrisMarsh82 commented on instat/Model/Output/clsOutputList.vb:
```diff
@@ -0,0 +1,81 @@
+    ''' </summary>
+    ''' <param name="outputElement"></param>
+    Public Sub MoveElementUp(outputElement As clsOutputElement)
+        If outputElement.Id > 1 Then
```
They do start at 1
outputElement.Id = filteredList.Output.Count + 1

ChrisMarsh82 commented:
Can't resolve the merge issue. Tried to use command line but it either wanted me to create a new pull request with all the changes duplicated or just didn't notice the conflict. - frmOutputWindow should just be deleted.

@dannyparsons  I fixed some of the issues however the refreshing when the dialog closes, scroll bars on the elements and font sizes don't seem to be an issue for me. I changed the dialogs running on 100% zoom to see if this makes any difference. Hopefully we can chat first thing tomorrow about it.

dannyparsons commented:
Could you maybe manually undo the changes to the conflicting files? It's just two and the changes are small. Then we could merge and redo the changes as needed.
